File: backend/src/onefive/api/files.py
"""
文件管理 API 路由

提供 115 云盘文件浏览、搜索、创建目录等接口。
"""
import asyncio
from fastapi import APIRouter
from pydantic import BaseModel
from ..models.schemas import ApiResponse
from ..services.file_service import get_file_service

router = APIRouter(prefix="/api/files", tags=["文件管理"])


class MkdirRequest(BaseModel):
    """创建目录请求"""
    name: str
    pid: int = 0


@router.get("/list", summary="列出目录内容")
async def list_files(
    cid: int = 0,
    limit: int = 100,
    offset: int = 0,
    order: str = "file_name",
    asc: int = 1,
):
    """列出指定目录下的文件和文件夹

    Args:
        cid: 目录 ID，0 表示根目录
        limit: 分页大小
        offset: 分页偏移
        order: 排序字段 (file_name/file_size/file_type/user_utime/user_ptime)
        asc: 排序方向 (1=升序 0=降序)
    """
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    result = await asyncio.to_thread(
        file_service.list_files, cid=cid, limit=limit, offset=offset, order=order, asc=asc
    )
    return ApiResponse(code=0, message="success", data=result)


@router.get("/info/{file_id}", summary="获取文件详情")
async def get_file_info(file_id: str):
    """获取文件或目录的详细信息"""
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    result = await asyncio.to_thread(file_service.get_file_info, file_id)
    return ApiResponse(code=0, message="success", data=result)


@router.post("/mkdir", summary="创建目录")
async def create_folder(req: MkdirRequest):
    """创建新目录"""
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    result = await asyncio.to_thread(file_service.create_folder, name=req.name, pid=req.pid)
    return ApiResponse(code=0, message="创建目录成功", data=result)


@router.get("/search", summary="搜索文件")
async def search_files(keyword: str, cid: int = 0):
    """搜索文件"""
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    result = await asyncio.to_thread(file_service.search_files, keyword=keyword, cid=cid)
    return ApiResponse(code=0, message="success", data=result)


# ==================== 批量操作 ====================

class MoveRequest(BaseModel):
    """移动文件请求"""
    file_ids: list[str]
    to_cid: str


class CopyRequest(BaseModel):
    """复制文件请求"""
    file_ids: list[str]
    to_cid: str


class DeleteRequest(BaseModel):
    """删除文件请求"""
    file_ids: list[str]


class RenameRequest(BaseModel):
    """重命名文件请求"""
    file_id: str
    new_name: str


@router.post("/move", summary="批量移动文件")
async def move_files(req: MoveRequest):
    """批量移动文件/目录到目标目录"""
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    await asyncio.to_thread(file_service.move_files, file_ids=req.file_ids, to_cid=req.to_cid)
    return ApiResponse(code=0, message="移动成功", data=None)


@router.post("/copy", summary="批量复制文件")
async def copy_files(req: CopyRequest):
    """批量复制文件/目录到目标目录"""
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    await asyncio.to_thread(file_service.copy_files, file_ids=req.file_ids, to_cid=req.to_cid)
    return ApiResponse(code=0, message="复制成功", data=None)


@router.post("/delete", summary="批量删除文件")
async def delete_files(req: DeleteRequest):
    """批量删除文件/目录（移入回收站）"""
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    await asyncio.to_thread(file_service.delete_files, file_ids=req.file_ids)
    return ApiResponse(code=0, message="删除成功", data=None)


@router.post("/rename", summary="重命名文件")
async def rename_file(req: RenameRequest):
    """重命名单个文件/目录"""
    file_service = get_file_service()
    # 115 API 网络请求为同步 I/O，放到线程池避免阻塞事件循环
    await asyncio.to_thread(file_service.rename_file, file_id=req.file_id, new_name=req.new_name)
    return ApiResponse(code=0, message="重命名成功", data=None)

File: frontend/src/api/files.ts
import api, { type ApiResult } from './index'

/** 文件/目录项 */
export interface FileItem {
  file_id: string
  name: string
  is_dir: boolean
  size: number
  file_type: number
  pick_code: string
  parent_id: string
  created_at: string
  updated_at: string
}

/** 文件列表响应 */
export interface FileListResponse {
  items: FileItem[]
  count: number
  offset: number
  limit: number
  parent_id: string
}

/** Open API 设置 */
export interface OpenApiSettings {
  enabled: boolean
  app_id: string
  token_valid: boolean
}

export const filesApi = {
  /** 列出目录内容 */
  listFiles(cid: string = '0', limit = 100, offset = 0, order = 'file_name', asc = 1): Promise<ApiResult<FileListResponse>> {
    return api.get('/files/list', { params: { cid, limit, offset, order, asc } })
  },

  /** 获取文件/目录详情 */
  getFileInfo(fileId: string): Promise<ApiResult<FileItem>> {
    return api.get(`/files/info/${fileId}`)
  },

  /** 创建目录 */
  createFolder(name: string, pid: string = '0'): Promise<ApiResult<any>> {
    return api.post('/files/mkdir', { name, pid: Number(pid) })
  },

  /** 搜索文件 */
  searchFiles(keyword: string, cid: string = '0'): Promise<ApiResult<FileListResponse>> {
    return api.get('/files/search', { params: { keyword, cid } })
  },

  /** 批量移动文件 */
  moveFiles(fileIds: string[], toCid: string): Promise<ApiResult<any>> {
    return api.post('/files/move', { file_ids: fileIds, to_cid: toCid })
  },

  /** 批量复制文件 */
  copyFiles(fileIds: string[], toCid: string): Promise<ApiResult<any>> {
    return api.post('/files/copy', { file_ids: fileIds, to_cid: toCid })
  },

  /** 批量删除文件 */
  deleteFiles(fileIds: string[]): Promise<ApiResult<any>> {
    return api.post('/files/delete', { file_ids: fileIds })
  },

  /** 重命名文件 */
  renameFile(fileId: string, newName: string): Promise<ApiResult<any>> {
    return api.post('/files/rename', { file_id: fileId, new_name: newName })
  },

  /** 获取 Open API 设置 */
  getOpenApiSettings(): Promise<ApiResult<OpenApiSettings>> {
    return api.get('/config/open-api')
  },

  /** 更新 Open API 设置 */
  updateOpenApiSettings(enabled: boolean, appId: string): Promise<ApiResult<any>> {
    return api.post('/config/open-api', { enabled, app_id: appId })
  },
}
